Add Reals set

rikardn opened the pull request:
Implementation of the set of real numbers: Reals.

isuruf commented on symengine/sets.cpp:
```diff
@@ -233,6 +233,46 @@ vec_basic Interval::get_args() const
 
+RCP<const Set> Reals::set_intersection(const RCP<const Set> &o) const
+{
+    return o;
```
`o` could have complex values right?

rikardn replied:
Thanks for pointing this out. I'll try to figure out how to do this. Ideas on how to handle complex numbers are welcome. I am thinking:

Interval, Empty - Must be real and can be handled as currently
FiniteSet - Can contain complex numbers. Could be handled directly
all others - Can contain complex numbers so create Intersection class

isuruf commented on symengine/sets.cpp:
```diff
@@ -233,6 +233,46 @@ vec_basic Interval::get_args() const
+
+RCP<const Set> Reals::set_union(const RCP<const Set> &o) const
+{
+    return reals();
```
Same here

isuruf reviewed (commented):
Thanks for sending this PR. Got a couple of questions

certik commented:
Thanks! There is a failure in https://travis-ci.org/github/symengine/symengine/jobs/727822895:
```
/home/travis/build/symengine/symengine/symengine/tests/printing/test_printing.cpp:692: FAILED:

  CHECK( latex(*l13) == "5 \\leq b \\wedge 2 \\leq a" )

with expansion:

  "2 \leq a \wedge 5 \leq b"

  ==

  "5 \leq b \wedge 2 \leq a"

/home/travis/build/symengine/symengine/symengine/tests/printing/test_printing.cpp:694: FAILED:

  CHECK( latex(*l14) == "b \\leq a \\wedge \\left(a \\neq c \\vee a = b\\right)" )

with expansion:

  "b \leq a \wedge \left(a = b \vee a \neq c\right)"

  ==

  "b \leq a \wedge \left(a \neq c \vee a = b\right)"
```

Not sure if it is related to this PR?

rikardn commented:
The test failure seems to be related to this PR. I reverted and ran the tests without fail. I will try to resolve it.

rikardn commented:
I see that I also missed adding Reals to is_a_Set. I'll do that.

rikardn commented:
I have 

* added ```Reals``` to ```is_a_Set```
* Fixed the ```set_union``` and ```set_intersection``` to handle complex numbers
* Fixed the failing test. At least for now. It seems as if the failing was due to reversed ordering of a set.

rikardn commented:
I see the code formatting issue. Is this fixed automatically or should I fix manually?

certik commented:
You have to manually apply the commands from the message.

isuruf commented on symengine/sets.cpp:
```diff
@@ -419,6 +471,19 @@ RCP<const Set> FiniteSet::set_union(const RCP<const Set> &o) const
+    if (is_a<Reals>(*o)) {
+        set_basic container;
+        for (const auto &elem : container_) {
+            if (is_a<Complex>(*elem)) {
```
Instead of checking if the value is a complex number here, we should check if the value is not a real number. For eg: symbols should be considered complex since we don't know what the value is going to be.

rikardn commented:
Thanks for your patience! I have updated again:

* Fixed the ```FiniteSet``` union and intersection as suggested
* Added a proper ```contains``` method which was forgotten

The intersection doesn't fully work because of missing functionality  in the ```set_intersection``` function. I don't know if you intend to have a separate class for intersections or if you want to solve it with unions and complements via deMorgan. In any case I don't know if anything else must be done about that in this PR.

isuruf commented on symengine/sets.cpp:
```diff
@@ -447,6 +529,35 @@ RCP<const Set> FiniteSet::set_intersection(const RCP<const Set> &o) const
+                others.insert(elem);
+            }
+        }
+        if (kept_reals.empty()) {
+            if (others.empty()) {
+                return emptyset();
+            } else {
+                return SymEngine::set_intersection(
+                    {reals(), finiteset(others)});
+            }
+        } else {
+            if (others.empty()) {
+                return finiteset(kept_reals);
+            } else {
+                others.insert(kept_reals.begin(), kept_reals.end());
+                return SymEngine::set_intersection(
+                    {reals(), finiteset(others)});
+            }
+        }
```
```suggestion
        others.insert(kept_reals.begin(), kept_reals.end());
        if (others.empty()) {
            return finiteset(kept_reals);
        } else {
            return SymEngine::set_intersection(
                    {reals(), finiteset(others)});
        }
```

isuruf commented on symengine/sets.cpp:
```diff
@@ -447,6 +529,35 @@ RCP<const Set> FiniteSet::set_intersection(const RCP<const Set> &o) const
+            if (others.empty()) {
+                return finiteset(kept_reals);
+            } else {
+                others.insert(kept_reals.begin(), kept_reals.end());
+                return SymEngine::set_intersection(
+                    {reals(), finiteset(others)});
```
```suggestion
                return set_union(finiteset(kept_reals), set_intersection(
                    {reals(), finiteset(others)}));
```

isuruf replied:
Above is another way to return it. Do you know what sympy does?

rikardn replied:
I tested:

```python
import sympy

x = sympy.Symbol('x')
finite = sympy.FiniteSet(sympy.Integer(1), sympy.Integer(2), sympy.I, x)
sympy.Reals.intersection(finite)
```

And got

```
Intersection(FiniteSet(1, 2, x), Reals)
```

Should be in line with my suggestion

isuruf replied:
Great. We'll go with your version

rikardn commented:
Now also the code coverage check passes. Are we good to go?

isuruf commented:
Thanks for the PR
